### DISCUSSION TITLE
ERROR failed to download from ... Double check that the given VERSION is valid

### DISCUSSION THREAD

**Discussion by LewisSpring** [itzg/docker-minecraft-bedrock-server#323] [Q&A] [answered]
Hi, this looks like a repeat of #115.
I can download the server via my web browser or via wget with no issue.
I've tried running as root and as user. 

I am using `LATEST` for the `VERSION` environment variable, but I have tried manually specifying `1.19.81.01`

DEBUG Output:
<details>

```$ docker logs -f minecraft-bedrock-1
DEBU[0000] Using /data to match uid and gid
DEBU[0000] Resolved UID=1000 from match path
DEBU[0000] Resolved GID=998 from match path
+ curlArgs=(-v)
++ id -a
++ ls -ld /data
+ echo 'DEBUG: running as uid=1000 gid=998 groups=998 with drwxrwsr-x+ 5 1000 998 4096 May  7 21:27 /data'
DEBUG: running as uid=1000 gid=998 groups=998 with drwxrwsr-x+ 5 1000 998 4096 May  7 21:27 /data
++ pwd
+ echo '       current directory is /data'
+ downloadPage=https://www.minecraft.net/en-us/download/server/bedrock
       current directory is /data
+ [[ TRUE != TRUE ]]
+ case ${VERSION^^} in
+ echo 'Looking up latest version...'
+ lookupVersion serverBedrockLinux
+ platform=serverBedrockLinux
+ for i in {1..3}
Looking up latest version...
++ restify --user-agent=itzg/minecraft-bedrock-server --headers 'accept-language:*' --attribute=data-platform=serverBedrockLinux https://www.minecraft.net/en-us/download/server/bedrock
++ jq -r '.[0].href'
+ DOWNLOAD_URL=https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip
+ [[ -n https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip ]]
+ break 2
+ [[ -z https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip ]]
+ [[ https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip =~ http.*/.*-(.*)\.zip ]]
+ VERSION=1.19.81.01
+ rm -f restify.err
+ '[' '!' -f bedrock_server-1.19.81.01 ']'
+ [[ ! -n https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip ]]
+ [[ /tmp != /tmp ]]
++ basename https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip
+ TMP_ZIP=/tmp/bedrock-server-1.19.81.01.zip
+ echo 'Downloading Bedrock server version 1.19.81.01 ...'
+ curl -v -o /tmp/bedrock-server-1.19.81.01.zip -fsSL https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip
Downloading Bedrock server version 1.19.81.01 ...
*   Trying 23.56.238.91:443...
* Connected to minecraft.azureedge.net (23.56.238.91) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [29 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [2627 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [80 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=US; ST=WA; L=Redmond; O=Microsoft Corporation; CN=*.azureedge.net
*  start date: Jul 27 23:19:41 2022 GMT
*  expire date: Jul 22 23:19:41 2023 GMT
*  subjectAltName: host "minecraft.azureedge.net" matched cert's "*.azureedge.net"
*  issuer: C=US; O=Microsoft Corporation; CN=Microsoft Azure ECC TLS Issuing CA 01
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
} [5 bytes data]
* Using Stream ID: 1 (easy handle 0x55e1d65472e0)
} [5 bytes data]
> GET /bin-linux/bedrock-server-1.19.81.01.zip HTTP/2
> Host: minecraft.azureedge.net
> user-agent: curl/7.74.0
> accept: */*
>
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [265 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [265 bytes data]
* old SSL session ID is stale, removing
{ [5 bytes data]
* Connection state changed (MAX_CONCURRENT_STREAMS == 100)!
} [5 bytes data]
< HTTP/2 200
< content-length: 81096423
< content-type: application/octet-stream
< content-md5: 8lQjOhDE61Cwz3MKa/40WA==
< last-modified: Thu, 27 Apr 2023 15:44:59 GMT
< accept-ranges: bytes
< etag: "0x8DB47365B6BA508"
< server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0
< x-ms-request-id: abc20167-f01e-0139-1723-793027000000
< x-ms-version: 2014-02-14
< x-ms-lease-status: unlocked
< x-ms-lease-state: available
< x-ms-blob-type: BlockBlob
< date: Sun, 07 May 2023 21:27:46 GMT
<
{ [5 bytes data]

* transfer closed with 60155656 bytes remaining to read
} [5 bytes data]
* stopped the pause stream!
* Connection #0 to host minecraft.azureedge.net left intact
curl: (18) transfer closed with 60155656 bytes remaining to read
+ echo 'ERROR failed to download from https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip'
+ echo '      Double check that the given VERSION is valid'
+ exit 2
ERROR failed to download from https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip
      Double check that the given VERSION is valid
```
</details>

Thanks!

**Comment by itzg** [comment]
Since it worked in the browser, then there is not a bug in the container startup. In that case it must be a networking issue in your docker installation or closer to the server host.

Where did you run wget?

The debug output shows the curl command that was used, so that's what you should have been re-trying.

Make sure to re-pull the image and debug/investigate your network while running:

```
docker run --rm --entrypoint curl itzg/minecraft-bedrock-server -v -o /tmp/bedrock-server-1.19.81.01.zip -fsSL https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip
```

**Reply by LewisSpring** [reply]
I ran wget in the docker host's terminal just as a sanity check.

I ran your command and got:
<details>

```
$ docker run --rm --entrypoint curl itzg/minecraft-bedrock-server -v -o /tmp/bedrock-server-1.19.81.01.zip -fsSL https://minecraft.azureedge.net/bin-linux/bedrock-server-1.19.81.01.zip
*   Trying 23.56.238.57:443...
* Connected to minecraft.azureedge.net (23.56.238.57) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [29 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [2627 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [80 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=US; ST=WA; L=Redmond; O=Microsoft Corporation; CN=*.azureedge.net
*  start date: Jul 27 23:19:41 2022 GMT
*  expire date: Jul 22 23:19:41 2023 GMT
*  subjectAltName: host "minecraft.azureedge.net" matched cert's "*.azureedge.net"
*  issuer: C=US; O=Microsoft Corporation; CN=Microsoft Azure ECC TLS Issuing CA 01
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
} [5 bytes data]
* Using Stream ID: 1 (easy handle 0x55b330cb12e0)
} [5 bytes data]
> GET /bin-linux/bedrock-server-1.19.81.01.zip HTTP/2
> Host: minecraft.azureedge.net
> user-agent: curl/7.74.0
> accept: */*
>
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [265 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [265 bytes data]
* old SSL session ID is stale, removing
{ [5 bytes data]
* Connection state changed (MAX_CONCURRENT_STREAMS == 100)!
} [5 bytes data]
< HTTP/2 200
< content-length: 81096423
< content-type: application/octet-stream
< content-md5: 8lQjOhDE61Cwz3MKa/40WA==
< last-modified: Thu, 27 Apr 2023 15:44:59 GMT
< accept-ranges: bytes
< etag: "0x8DB47365B6BA508"
< server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0
< x-ms-request-id: f0e509fa-701e-0167-3763-7ec324000000
< x-ms-version: 2014-02-14
< x-ms-lease-status: unlocked
< x-ms-lease-state: available
< x-ms-blob-type: BlockBlob
< date: Mon, 08 May 2023 09:26:08 GMT
<
{ [5 bytes data]
* transfer closed with 62415616 bytes remaining to read
} [5 bytes data]
* stopped the pause stream!
* Connection #0 to host minecraft.azureedge.net left intact
curl: (18) transfer closed with 62415616 bytes remaining to read
```

</details>

I tried using Host and Bridge with that command.
I have other containers running on this system and they're not having issues with internet connectivity.
Please let me know if there's anything else to try.
Thanks

**Reply by LewisSpring** [reply]
Is there a workaround method to do a manual download? (https://github.com/itzg/docker-minecraft-bedrock-server/issues/121)
If I place the download in the container's `/tmp/` will it continue without attempting to download?

**Reply by itzg** [reply, marked as answer]
> Is there a workaround method to do a manual download? (#121)

Not yet. That's why there is the open issue you linked. I will gladly accept a PR to add that support -- download issues are extremely rare and not caused by this image, so I'm afraid it goes way at the bottom of my todo list.

**Reply by LewisSpring** [reply]
That's fair enough.
I'd tried a lot of things and put it to the side, but a power outage today has fixed it. Server on UPS and so is local DNS. No clue where the issue was actually but I'm glad it works now. 

I apologise for the non-solution, but thank you for your time!

**Comment by tuknet2018** [comment]
I have the same issue. I’ve tried using Podman and Docker on different servers, and while all the containers have internet access, the problem persists: curl cannot download the .zip file. I attempted to change the entry point to /bin/sh and manually download the .zip with the agent options for curl. If I try curl without the -A parameter, it doesn't work. Downloading the file normally via the endpoint seems impossible.

The problem is the same, included in new "installation" 
`
[root@podman /]# podman logs recursing_sammet
DEBU[0000] Using /data to match uid and gid
DEBU[0000] Resolved UID=0 from match path
DEBU[0000] Resolved GID=0 from match path
Looking up latest version...
Downloading Bedrock server version 1.21.31.04 ...
curl: (6) getaddrinfo() thread failed to start
ERROR failed to download from https://www.minecraft.net/bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip
      Double check that the given VERSION is valid`

**Reply by jurgenweber** [reply]
Yeah, I am getting this too; odd;

```
curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)
ERROR failed to download from https://www.minecraft.net/bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip
      Double check that the given VERSION is valid
DEBU[0000] Using /data to match uid and gid
DEBU[0000] Resolved UID=0 from match path
DEBU[0000] Resolved GID=2000 from match path
INFO[0000] Skipping uid/gid change since current user is not root
Looking up latest version...
Downloading Bedrock server version 1.21.31.04 ...
curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)
```

**Reply by jurgenweber** [reply]
investigating further; if I run the server on my k8s cluster (using the helm chart) in restore mode; 

```
$ curl -v -o /tmp/bedrock-server-1.21.31.04.zip -fsSL https://www.minecraft.net/bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip
*   Trying 203.87.122.19:443...
* Connected to www.minecraft.net (203.87.122.19) port 443 (#0)
* ALPN: offers h2,http/1.1
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [29 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [2427 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [80 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=WA; L=Redmond; O=Microsoft Corporation; CN=*.minecraft.net
*  start date: May 29 20:28:20 2024 GMT
*  expire date: May 24 20:28:20 2025 GMT
*  subjectAltName: host "www.minecraft.net" matched cert's "*.minecraft.net"
*  issuer: C=US; O=Microsoft Corporation; CN=Microsoft Azure ECC TLS Issuing CA 04
*  SSL certificate verify ok.
} [5 bytes data]
* using HTTP/2
* h2h3 [:method: GET]
* h2h3 [:path: /bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip]
* h2h3 [:scheme: https]
* h2h3 [:authority: www.minecraft.net]
* h2h3 [user-agent: curl/7.88.1]
* h2h3 [accept: */*]
* Using Stream ID: 1 (easy handle 0x55739b790c80)
} [5 bytes data]
> GET /bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip HTTP/2
> Host: www.minecraft.net
> user-agent: curl/7.88.1
> accept: */*
>
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [265 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [265 bytes data]
* old SSL session ID is stale, removing
{ [5 bytes data]
* HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)
* Connection #0 to host www.minecraft.net left intact
curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)
```

it is the same if I run it locally on my macbook pro; and pretty much every machine I try.. :\

**Reply by jurgenweber** [reply]
ok, figured it out. It seems Microsoft is banning the curl user agents; refernce; https://github.com/curl/curl/issues/12114

this worked fine; 

```
$ curl -A "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36" -v -o /tmp/bedrock-server-1.21.31.04.zip https://www.minecraft.net/bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host www.minecraft.net:443 was resolved.
* IPv6: (none)
* IPv4: 23.46.179.50, 23.46.179.17
*   Trying 23.46.179.50:443...
* Connected to www.minecraft.net (23.46.179.50) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
} [322 bytes data]
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* (304) (IN), TLS handshake, Unknown (8):
{ [29 bytes data]
* (304) (IN), TLS handshake, Certificate (11):
{ [2427 bytes data]
* (304) (IN), TLS handshake, CERT verify (15):
{ [78 bytes data]
* (304) (IN), TLS handshake, Finished (20):
{ [36 bytes data]
* (304) (OUT), TLS handshake, Finished (20):
} [36 bytes data]
* SSL connection using TLSv1.3 / AEAD-CHACHA20-POLY1305-SHA256 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=WA; L=Redmond; O=Microsoft Corporation; CN=*.minecraft.net
*  start date: May 29 20:28:20 2024 GMT
*  expire date: May 24 20:28:20 2025 GMT
*  subjectAltName: host "www.minecraft.net" matched cert's "*.minecraft.net"
*  issuer: C=US; O=Microsoft Corporation; CN=Microsoft Azure ECC TLS Issuing CA 04
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://www.minecraft.net/bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: www.minecraft.net]
* [HTTP/2] [1] [:path: /bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip]
* [HTTP/2] [1] [user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36]
* [HTTP/2] [1] [accept: */*]
> GET /bedrockdedicatedserver/bin-linux/bedrock-server-1.21.31.04.zip HTTP/2
> Host: www.minecraft.net
> User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36
> Accept: */*
>
* Request completely sent off
< HTTP/2 200
< content-length: 56027291
< content-type: application/x-zip-compressed
< content-md5: gPPO/pbSYdvJnG61s5DuDA==
< last-modified: Mon, 30 Sep 2024 15:45:49 GMT
< accept-ranges: bytes
< etag: "0x8DCE166F5659C92"
< server: Windows-Azure-Web/1.0 Microsoft-HTTPAPI/2.0
< x-ms-request-id: e4205f8a-201e-003c-0850-13e158000000
< x-ms-version: 2018-03-28
< cache-control: max-age=127890
< expires: Sun, 13 Oct 2024 23:04:15 GMT
< date: Sat, 12 Oct 2024 11:32:45 GMT
< server-timing: cdn-cache; desc=HIT
< server-timing: edge; dur=5
< server-timing: ak_p; desc="1728732765932_388936461_65033234_505_11049_11_15_15";dur=1
<
{ [16375 bytes data]
100 53.4M  100 53.4M    0     0   9.8M      0  0:00:05  0:00:05 --:--:--  9.9M
* Connection #0 to host www.minecraft.net left intact
```

Maybe this is known, since in the entrypoint you are setting the user-agent;

-A "itzg/minecraft-bedrock-server"

which also works for me locally...

**Reply by jurgenweber** [reply]
Right, got it.. this was discovered 3 weeks ago; https://github.com/itzg/docker-minecraft-bedrock-server/pull/459

but the default imagePullPolicy in the chart is 'IfNotPresent' and since the lag is latest, I am on the old image still; https://github.com/itzg/minecraft-server-charts/blob/master/charts/minecraft-bedrock/values.yaml.... becuase I ran it up in debug and I don't see the header being set.

If you are going to set the default tag to 'latest' then this (https://github.com/itzg/minecraft-server-charts/blob/master/charts/minecraft-bedrock/values.yaml#L5C3-L5C27) should be 'Always'.

I am working again.

**Comment by bgunn925** [comment]
Hi, I am experiencing the same issue but the conversation here is a bit over my head, would someone be so kind as to ELI5 how to implement the above solution? Thank you!

**Reply by itzg** [reply]
You didn't share your logs so it's not apparent you have the same issue. 

Re-pull the image and it will fix an unrelated download issue that was caused by Microsoft recently.

**Reply by bgunn925** [reply]
I'm using Portainer and had to figure out how to re-pull the image but that seemed to work, thank you
